File: javascript/lasagna-master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

'use strict';

/**
 * @param {number} [timeLeft]
 */
function cookingStatus(timeLeft) {

    switch (timeLeft) {
        case undefined:
            return "You forgot to set the timer."
        case 0:
            return "Lasagna is done."
        default:
            return "Not done, please wait.";
    }
}

/**
 * @param {string[]} layers
 * @param {number} prepTime
 */
function preparationTime(layers, prepTime = 2) {

    // do JS devs test for undefined and null?
    return layers.length * prepTime;
}

/**
 * 
 * @param {string[]} layers 
 */
function quantities(layers) {

    let result = {
        noodles: 0,
        sauce: 0,
    };

    layers.forEach(layer => {

        if (layer === 'noodles') {
            result.noodles += 50;
        } else if (layer === 'sauce') {
            result.sauce += 0.2;
        }
    });

    return result;
}

/**
 * @param {string[]} friendsList
 * @param {string[]} myList
 */
function addSecretIngredient(friendsList, myList) {

    // assuming there is at least one item in `friendsList`
    myList.push(friendsList.at(-1));
}

/**
 * @param {object} recipe
 * @param {number} scale
 */
function scaleRecipe(recipe, scale) {

    // we don't need a deep copy
    let newRecipe = {...recipe};

    // determine single value and multiply by scale for each object property
    // assuming the object passed in is correct
    for (const key in newRecipe) {
        newRecipe[key] = (newRecipe[key] / 2) * scale;
    }

    return newRecipe;
}

export {cookingStatus, preparationTime, quantities, scaleRecipe, addSecretIngredient}

File: go/lasagna-master/lasagna_master.go
package lasagna

const noodlesPerLayer = 50
const saucePerLayer = 0.2

// PreparationTime estimate for the total preparation time based on the number of layers and time per layer
func PreparationTime(layers []string, time int) int {

	if time < 1 {
		time = 2
	}

	return len(layers) * time
}

// Quantities determine the quantity of noodles and sauce needed
func Quantities(layers []string) (int, float64) {

	var noodleCount int
	var sauceCount int

	for _, val := range layers {

		if val == "noodles" {
			noodleCount++
		}

		if val == "sauce" {
			sauceCount++
		}
	}

	noodles := noodleCount * noodlesPerLayer
	sauce := float64(sauceCount) * saucePerLayer

	return noodles, sauce
}

// AddSecretIngredient
func AddSecretIngredient(friendsList []string, myList []string) {

	secret := friendsList[len(friendsList)-1]
	myList[len(myList)-1] = secret
}

// ScaleRecipe
func ScaleRecipe(quantities []float64, portions int) []float64 {

	scale := float64(portions) / 2.0
	newQuantities := make([]float64, len(quantities))

	for idx, _ := range newQuantities {
		newQuantities[idx] = quantities[idx] * float64(scale)
	}

	return newQuantities
}
